keyboards/ploopyco/trackball_thumb/trackball_thumb.c

```diff
@@ -0,0 +1,229 @@
+    }
+
+    // Don't scroll if the middle button is depressed.
+    if (is_scroll_clicked) {
+#ifndef IGNORE_SCROLL_CLICK
+        return false;
+#endif
+    }
+
+    last_scroll = timer_read();
+    uint16_t p1 = adc_read(OPT_ENC1_MUX);
+    uint16_t p2 = adc_read(OPT_ENC2_MUX);
+    if (debug_encoder) dprintf("OPT1: %d, OPT2: %d\n", p1, p2);
+
+    int dir = opt_encoder_handler(p1, p2);
+
+    if (dir == 0) return false;
+    ;
+    encoder_update_kb(0, dir == 1);
+    return true;
+}
+
+report_mouse_t pointing_device_task_kb(report_mouse_t mouse_report) {
+    if (is_drag_scroll) {
+        mouse_report.h = mouse_report.x;
+#ifdef PLOOPY_DRAGSCROLL_INVERT
+        // Invert vertical scroll direction
+        mouse_report.v = -mouse_report.y;
+#else
+        mouse_report.v = mouse_report.y;
+#endif
+        mouse_report.x = 0;
+        mouse_report.y = 0;
+    }
+
+    return pointing_device_task_user(mouse_report);
+}
+
+bool process_record_kb(uint16_t keycode, keyrecord_t* record) {
+    // Update Timer to prevent accidental scrolls
+    if ((record->event.key.col == 1) && (record->event.key.row == 0)) {
+        last_mid_click    = timer_read();
+        is_scroll_clicked = record->event.pressed;
+    }
+
+    if (!process_record_user(keycode, record)) {
+        return false;
+    }
+
+    if (keycode == DPI_CONFIG && record->event.pressed) {
+        keyboard_config.dpi_config = (keyboard_config.dpi_config + 1) % DPI_OPTION_SIZE;
+        eeconfig_update_kb(keyboard_config.raw);
+        pointing_device_set_cpi(dpi_array[keyboard_config.dpi_config]);
+    }
+
+    if (keycode == DRAG_SCROLL) {
+#ifndef PLOOPY_DRAGSCROLL_MOMENTARY
+        if (record->event.pressed)
+#endif
+        {
+            is_drag_scroll ^= 1;
+        }
+#ifdef PLOOPY_DRAGSCROLL_FIXED
+        pointing_device_set_cpi(is_drag_scroll ? PLOOPY_DRAGSCROLL_DPI : dpi_array[keyboard_config.dpi_config]);
+#else
+        pointing_device_set_cpi(is_drag_scroll ? (dpi_array[keyboard_config.dpi_config] * PLOOPY_DRAGSCROLL_MULTIPLIER) : dpi_array[keyboard_config.dpi_config]);
+#endif
+    }
+
+/* If Mousekeys is disabled, then use handle the mouse button
+ * keycodes.  This makes things simpler, and allows usage of
+ * the keycodes in a consistent manner.  But only do this if
+ * Mousekeys is not enable, so it's not handled twice.
+ */
+#ifndef MOUSEKEY_ENABLE
+    if (IS_MOUSEKEY_BUTTON(keycode)) {
+        report_mouse_t currentReport = pointing_device_get_report();
+        if (record->event.pressed) {
+            currentReport.buttons |= 1 << (keycode - KC_MS_BTN1);
+        } else {
+            currentReport.buttons &= ~(1 << (keycode - KC_MS_BTN1));
+        }
+        pointing_device_set_report(currentReport);
+        pointing_device_send();
+    }
+#endif
+
+    return true;
+}
+
+// Hardware Setup
+void keyboard_pre_init_kb(void) {
+    // debug_enable  = true;
+    // debug_matrix  = true;
+    // debug_mouse   = true;
+    // debug_encoder = true;
+
```

Comment:
These should be set in `keyboard_post_init_kb`; `debug_enable`, at least, would be overridden in `keyboard_init`: https://github.com/qmk/qmk_firmware/blob/cf88d956130ef847677cf4defd1f21d4e72c677f/quantum/keyboard.c#L415

Reply:
In this case, it really shouldn't matter.  These default to false, normally, and are loaded from eeprom.

Mostly, they're here for reference.